File: elixir/plugins/azuresecrets.ex
# Azure Key Vault, as a sekreto plugin.
#
# A port of typescript/plugins/azuresecrets.ts, which is canonical.

defmodule Sekreto.Plugins.Azuresecrets do
  @moduledoc """
  The `azuresecrets` provider kind.

  A plugin module: nothing under `src/` names it, and a chain that does not
  name this kind never loads it. See docs/design/plugin-providers.md.
  """

  alias Sekreto.Cell
  alias Sekreto.Error
  alias Sekreto.Json
  alias Sekreto.Plugins.Http
  alias Sekreto.Plugins.Httpjson
  alias Sekreto.ProviderSpec
  alias Sekreto.Providers

  # The Key Vault audience an Azure token is minted for.
  @resource "https://vault.azure.net"

  @doc """
  The `azuresecrets` provider kind, as a voxgig/plugin definition.

  The calling project passes it to the constructor, and only then can a
  chain name the kind:

      Sekreto.new(chain, plugins: [Sekreto.Plugins.Azuresecrets.azuresecrets()])
  """
  def azuresecrets, do: Providers.providerplugin("azuresecrets", &fromspec/1)

  @doc """
  Azure Key Vault.

  `api.token` reads secret `api-token` (dots flattened to `-`; Key Vault
  names allow nothing else), current version. The token comes from config,
  then a client-credentials login when tenant/clientid/clientsecret are
  given, then the IMDS managed-identity endpoint - so on Azure's own
  platform no credential configuration is needed.
  """
  def azuresecrets(opts) do
    cell = Cell.new(%{token: "", renewat: Httpjson.never()})

    login = fn ->
      cond do
        "" != opts.token ->
          Cell.put(cell, %{token: opts.token, renewat: Httpjson.never()})

        "" != opts.tenant and "" != opts.clientid and "" != opts.clientsecret ->
          useloginaddr = Providers.first([opts.loginaddr, "https://login.microsoftonline.com"])
          Providers.checkaddr(useloginaddr)

          url = Httpjson.trimslash(useloginaddr) <> "/" <> opts.tenant <> "/oauth2/v2.0/token"

          form =
            "grant_type=client_credentials&client_id=" <>
              Http.uriescape(opts.clientid) <>
              "&client_secret=" <>
              Http.uriescape(opts.clientsecret) <>
              "&scope=" <> Http.uriescape(@resource <> "/.default")

          res =
            Httpjson.fetchjson(
              "POST",
              url,
              [{"content-type", "application/x-www-form-urlencoded"}],
              form
            )

          got = Json.text(Json.dig(res.body, ["access_token"]))

          if 200 != res.status or :none == got or "" == got do
            raise Error, message: "sekreto: azure login failed: " <> Httpjson.tostr(res.status)
          end

          renewat = Httpjson.renewtime(Json.dig(res.body, ["expires_in"]))
          Cell.put(cell, %{token: got, renewat: renewat})

        true ->
          imds =
            Httpjson.trimslash(Providers.first([opts.imdsaddr, "http://169.254.169.254"])) <>
              "/metadata/identity/oauth2/token?api-version=2018-02-01&resource=" <>
              Http.uriescape(@resource)

          res = Httpjson.fetchjson("GET", imds, [{"Metadata", "true"}])
          got = Json.text(Json.dig(res.body, ["access_token"]))

          if 200 != res.status or :none == got or "" == got do
            raise Error,
              message: "sekreto: azure: no token, no client credentials, and IMDS did not answer"
          end

          # IMDS sends expires_in as a STRING, unlike everyone else.
          renewat = Httpjson.renewtime(Json.dig(res.body, ["expires_in"]))
          Cell.put(cell, %{token: got, renewat: renewat})
      end
    end

    %{
      lookup: fn name ->
        if "" == opts.vault, do: raise(Error, message: "sekreto: azure: no vault")

        # Only an explicit scheme is a URL; a vault NAMED httpvault must
        # still become https://httpvault.vault.azure.net.
        vaulturl =
          if String.starts_with?(opts.vault, "http://") or
               String.starts_with?(opts.vault, "https://") do
            opts.vault
          else
            "https://" <> opts.vault <> ".vault.azure.net"
          end

        Providers.checkaddr(vaulturl)

        if Httpjson.expired?(cell), do: login.()

        url =
          Httpjson.trimslash(vaulturl) <>
            "/secrets/" <>
            Sekreto.flatname(name, "-") <>
            "?api-version=" <> Providers.first([opts.apiversion, "7.4"])

        res =
          Httpjson.fetchjson("GET", url, [{"authorization", "Bearer " <> Cell.get(cell).token}])

        cond do
          404 == res.status ->
            nil

          200 != res.status ->
            raise Error,
              message:
                "sekreto: azure error: " <> Httpjson.tostr(res.status) <> ": " <> Http.bare(url)

          true ->
            Httpjson.nonone(Json.text(Json.dig(res.body, ["value"])))
        end
      end,
      describe: fn -> "azuresecrets:" <> opts.vault end
    }
  end

  defp fromspec(%ProviderSpec{} = spec) do
    azuresecrets(%{
      vault: spec.vault,
      token: spec.token,
      tenant: spec.tenant,
      clientid: spec.clientid,
      clientsecret: spec.clientsecret,
      loginaddr: spec.loginaddr,
      imdsaddr: spec.imdsaddr,
      apiversion: spec.apiversion
    })
  end
end

File: haskell/plugins/Azuresecrets.hs
-- | Azure Key Vault, as a plugin.
--
-- A port of typescript/plugins/azure.ts, which is canonical.

module Azuresecrets
  ( azuresecrets,
    azuresecretsprovider,
  )
where

import Control.Monad (when)
import Data.IORef (newIORef, writeIORef)
import Data.List (isPrefixOf)
import Data.Maybe (fromMaybe)
import Defs (Definition)
import Httpjson (Answer (..), currenttoken, fetchjson, nakedurl, never, renewtime)
import qualified Json
import Names (flatname)
import Provider (Provider (..), forced)
import Providers
  ( ProviderSpec (..),
    checkaddr,
    fail',
    first,
    providerplugin,
    trimslash,
  )
import Sigv4 (uriescape)


-- | The Key Vault audience an Azure token is minted for.
azureresource :: String
azureresource = "https://vault.azure.net"

-- | Azure Key Vault.
--
-- @api.token@ reads secret @api-token@ (dots flattened to @-@; Key Vault
-- names allow nothing else), current version. The token comes from
-- config, then a client-credentials login when tenant/clientid/
-- clientsecret are given, then the IMDS managed-identity endpoint.
--
-- As with GCP, the IMDS call is plain http to a link-local host by
-- platform design and carries no credential; the login and vault
-- addresses are `checkaddr`-guarded.
azuresecretsprovider :: ProviderSpec -> IO Provider
azuresecretsprovider spec = do
  livetoken <- newIORef Nothing
  renewat <- newIORef never

  let login
        | not (null (spectoken spec)) = pure (spectoken spec)
        | not (null (spectenant spec))
            && not (null (specclientid spec))
            && not (null (specclientsecret spec)) = do
            let useloginaddr = first [specloginaddr spec, "https://login.microsoftonline.com"]
            checkaddr useloginaddr

            let url = trimslash useloginaddr ++ "/" ++ spectenant spec ++ "/oauth2/v2.0/token"
                form =
                  "grant_type=client_credentials&client_id="
                    ++ uriescape (specclientid spec)
                    ++ "&client_secret="
                    ++ uriescape (specclientsecret spec)
                    ++ "&scope="
                    ++ uriescape (azureresource ++ "/.default")

            res <-
              fetchjson
                "POST"
                url
                [("content-type", "application/x-www-form-urlencoded")]
                (Just form)

            let got = Json.text (Json.dig (ansbody res) ["access_token"])

            when (200 /= ansstatus res || maybe True null got) $
              fail' ("sekreto: azure login failed: " ++ show (ansstatus res))

            renewtime (Json.dig (ansbody res) ["expires_in"]) >>= writeIORef renewat
            pure (fromMaybe "" got)
        | otherwise = do
            let imds =
                  trimslash (first [specimdsaddr spec, "http://169.254.169.254"])
                    ++ "/metadata/identity/oauth2/token?api-version=2018-02-01&resource="
                    ++ uriescape azureresource

            res <- fetchjson "GET" imds [("Metadata", "true")] Nothing

            let got = Json.text (Json.dig (ansbody res) ["access_token"])

            when (200 /= ansstatus res || maybe True null got) $
              fail' "sekreto: azure: no token, no client credentials, and IMDS did not answer"

            renewtime (Json.dig (ansbody res) ["expires_in"]) >>= writeIORef renewat
            pure (fromMaybe "" got)

  pure
    Provider
      { lookupsecret = \name -> do
          when (null (specvault spec)) $ fail' "sekreto: azure: no vault"

          -- Only an explicit scheme is a URL; a vault NAMED httpvault must
          -- still become https://httpvault.vault.azure.net.
          let usevault = specvault spec
              vaulturl =
                if "http://" `isPrefixOf` usevault || "https://" `isPrefixOf` usevault
                  then usevault
                  else "https://" ++ usevault ++ ".vault.azure.net"

          checkaddr vaulturl

          token <- currenttoken livetoken renewat login

          flat <- forced (flatname name "-")

          let url =
                trimslash vaulturl
                  ++ "/secrets/"
                  ++ flat
                  ++ "?api-version="
                  ++ first [specapiversion spec, "7.4"]

          res <- fetchjson "GET" url [("authorization", "Bearer " ++ token)] Nothing

          if 404 == ansstatus res
            then pure Nothing
            else
              if 200 /= ansstatus res
                then
                  fail'
                    ("sekreto: azure error: " ++ show (ansstatus res) ++ ": " ++ nakedurl url)
                else pure (Json.text (Json.dig (ansbody res) ["value"])),
        describe = "azuresecrets:" ++ specvault spec
      }


-- | @azuresecrets@, as a voxgig/plugin definition.
azuresecrets :: Definition
azuresecrets = providerplugin "azuresecrets" azuresecretsprovider
